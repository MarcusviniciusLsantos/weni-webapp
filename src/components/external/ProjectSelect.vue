<template>
  <UnnnicFormElement
    class="unnnic-form-element"
    :label="$t('SIDEBAR.PROJECT')"
    fixedLabel
    size="sm"
  >
    <UnnnicSelectSmart
      v-if="canCreateProject"
      :disabled="projects.status === 'loading'"
      :key="projects.data.length"
      size="sm"
      :value="
        projects.status === 'loading'
          ? []
          : [
              projects.data
                .map(({ name, uuid }) => ({
                  value: uuid,
                  label: name,
                }))
                .find(({ value }) => value === currentProject.uuid) || null,
            ]
      "
      @input="changeProject($event[0].value)"
      :options="
        [
          {
            value: '',
            label: $t('loading'),
          },
          {
            value: 'create',
            label: $t('NAVBAR.PROJECT_CREATE'),
          },
          {
            value: 'see all',
            label: $t('NAVBAR.ALL_PROJECTS'),
          },
        ].concat(
          projects.data.map(({ name, uuid }) => ({
            value: uuid,
            label: name,
          })),
        )
      "
      orderedByIndex
    />

    <UnnnicInput
      v-else
      size="sm"
      :value="currentProject.name"
      iconRight="arrow-button-down-1"
      disabled
    ></UnnnicInput>
  </UnnnicFormElement>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';
import { ORG_ROLE_ADMIN, ORG_ROLE_CONTRIBUTOR } from '../orgs/orgListItem.vue';

export default {
  name: 'ProjectSelect',
  props: {
    org: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      page: 0,
    };
  },
  computed: {
    ...mapGetters(['currentProject']),

    projects() {
      return this.$store.state.Project.projects.find(
        ({ orgUuid }) => orgUuid === this.org.uuid,
      );
    },

    canCreateProject() {
      return [ORG_ROLE_ADMIN, ORG_ROLE_CONTRIBUTOR].includes(
        this.$store.getters.org.authorization.role,
      );
    },

    orgName() {
      if (!this.org) return null;
      return this.org.name;
    },
  },

  watch: {
    'projects.status': {
      immediate: true,
      handler(status) {
        if (status !== 'complete') {
          this.$store.dispatch('loadProjects', {
            orgUuid: this.org.uuid,
            ordering: '-created_at',
          });
        }
      },
    },
  },

  methods: {
    ...mapActions(['setCurrentProject']),

    allProjects() {
      this.$router.push({
        name: 'projects',
        params: {
          orgUuid: this.org.uuid,
        },
      });
    },

    changeProject(uuid) {
      if (uuid === this.currentProject.uuid) {
        return;
      }

      if (uuid === 'create') {
        this.$router.push({
          name: 'project_create',
          params: {
            orgUuid: this.org.uuid,
          },
        });
        return;
      }

      if (uuid === 'see all') {
        this.allProjects();
        return;
      }

      const project = this.projects.data.find(
        (project) => project.uuid === uuid,
      );

      if (!project) return;

      this.setCurrentProject(project);

      const insightsIframe = document.querySelector('iframe[name="insights"]');

      insightsIframe.contentWindow.postMessage(
        { event: 'setProject', projectUuid: project.uuid },
        '*',
      );

      this.$router.push({
        params: {
          projectUuid: project.uuid,
        },
      });
    },
  },
};
</script>

<style lang="scss" scoped>
.unnnic-form-element :deep(.label) {
  z-index: 2;
}
</style>
