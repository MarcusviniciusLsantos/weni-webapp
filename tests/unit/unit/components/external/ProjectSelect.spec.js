import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import ProjectSelect from '@/components/external/ProjectSelect.vue';
import i18n from '@/utils/plugins/i18n';

import Vuex from 'vuex';

import { org, project } from '../../../__mocks__';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('ProjectSelect.vue', () => {
  let wrapper;
  let getters;
  let state;
  let store;

  beforeEach(() => {
    state = {
      Project: {
        projects: [],
      },
    };
    getters = {
      currentProject() {
        return project;
      },
      org() {
        return org;
      },
    };
    store = new Vuex.Store({ state, getters });
    wrapper = shallowMount(ProjectSelect, {
      localVue,
      i18n,
      propsData: {
        org,
      },
      store,
      mocks: {
        $t: () => 'some specific text',
      },
      stubs: {
        unnnicAutoComplete: true,
        projectSelect: true,
        unnnicToolTip: true,
        unnnicIconSvg: true,
        RouterLink: RouterLinkStub,
        unnnicLanguageSelect: true,
        unnnicDropdown: true,
        avatar: true,
        unnnicSelect: true,
      },
    });
  });

  it('renders a snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
