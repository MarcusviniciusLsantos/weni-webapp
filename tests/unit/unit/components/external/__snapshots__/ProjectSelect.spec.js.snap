// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectSelect.vue > renders a snapshot 1`] = `
<unnnicformelement
  class="unnnic-form-element"
  data-v-14b459b5=""
  fixedlabel=""
  label="some specific text"
  size="sm"
>
  <unnnicinput
    data-v-14b459b5=""
    disabled="disabled"
    iconright="arrow-button-down-1"
    size="sm"
    value="Teste proj"
  />
</unnnicformelement>
`;
